[Doc] [cc] Doc the workflow for exporting Taichi kernels to C99 source (#1756)
taichi-dev/taichi: docs/export_kernels.rst

Export Taichi kernels to C source
=================================

The C backend of Taichi allows you to **export Taichi kernels to C source**.

> The exported Taichi program consists purely of C99 compatible sources and does not require Python. This allows you use the exported sources in a C/C++ project, or even call them from Javascript via Emscripten.

Reviewer: ```suggestion
The exported Taichi program consists purely of C99-compatible code and does not require Python. This allows you to use the exported code in a C/C++ project, or even to further compile it to Javascript/Web Assembly via Emscripten.
```

> The exported Taichi program consists purely of C99 compatible sources and does not require Python. Instead, you can use the exported sources in a C/C++ project, or even by calling them in Javascript via Empscripten.

Reviewer 2: ```suggestion
The exported Taichi program consists purely of C99 compatible sources and does not require Python. This lets you use the exported sources in a C/C++ project, or even call them from Javascript via Emscripten.
```

Author: ```suggestion
The exported Taichi program consists purely of C99 compatible sources and does not require Python. This allows you use the exported sources in a C/C++ project, or even call them from Javascript via Emscripten.
```

> Each C function corresponds to a Taichi kernel.

Reviewer: ```suggestion
Each C function corresponds to one Taichi kernel.
```

> For example, ``Tk_init_c6_0()`` may corresponds to ``init()`` in ``mpm88.py``.

Reviewer 2: ```suggestion
For example, ``Tk_init_c6_0()`` may correspond to ``init()`` in ``mpm88.py``.
```

> Only the source is needed when copying, all the required Taichi runtimes are
> included in that single C source file.

Reviewer 2: ```suggestion
Only the exported source is needed when copying or moving around the code, all of the required Taichi runtimes are
included in that single C source file.
```

> Only the exported source is needed when copying or moving around the code, all of the required Taichi runtimes are
> included in that single C source file.

Reviewer: ```suggestion
The exported C code is self-contained for portability. Required Taichi runtime functions are included in the code.
```

> For example, this also allows commercial users to distribute their Taichi program in
> binary format by linking this file with their project.

Reviewer: ```suggestion
For example, this allows programmers to distribute Taichi programs in
a binary format, by compiling and linking exported C code to their project.
```

.. warning::

>     Currently, this feature is only officially supported on the C backend which is only released on **Linux** at this moment.

Reviewer 2: ```suggestion
    Currently, this feature is only officially supported on the C backend which is only released on **Linux**.
```

Author: We'll soon release on OS X and Windows later.

>     Currently, this feature is only officially supported on the C backend which is only released on **Linux** at this moment.

Reviewer: ```suggestion
    Currently, this feature is only officially supported on the C backend on Linux. In the future, we will support OS X and Windows.
```

> The workflow of exporting
> ----------------------

Reviewer 2: ```suggestion
The workflow of exporting
-------------------------
```

Use ``ti.core.start_recording`` in the Taichi program you want to export.

> Suppose you want to export the `examples/mpm88.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm88.py>`_, here are the workflow:

Reviewer 2: ```suggestion
Suppose you want to export `examples/mpm88.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm88.py>`_, here is the workflow:
```

Export YAML
+++++++++++

> First, modifying the ``mpm88.py`` as shown below:

Reviewer 2: ```suggestion
First, modify ``mpm88.py`` as shown below:
```

.. code-block:: python

    import taichi as ti

    ti.core.start_recording('mpm88.yml')
    ti.init(arch=ti.cc)

    ... # your program

> And run it. Close the GUI window once particles are shown up correctly.

Reviewer: ```suggestion
Then please execute ``mpm88.py``. Close the GUI window once particles are shown up correctly.
```

This will save all the kernels in ``mpm88.py`` to ``mpm88.yml``:

.. code-block:: yaml

   - action: "compile_kernel"
      kernel_name: "init_c6_0"
      kernel_source: "void Tk_init_c6_0(struct Ti_Context *ti_ctx) {\n  for (Ti_i32 tmp0 = 0; tmp0 < 8192...\n"
    - action: "launch_kernel"
      kernel_name: "init_c6_0"
    ...

.. note::

    Equivalently, you may also specify these two arguments from environment
    variables on Unix-like system:

    .. code-block:: bash

        TI_ARCH=cc TI_ACTION_RECORD=mpm88.yml python mpm88.py

> Compose YAML into a single C file
> +++++++++++++++++++++++++++++++

Reviewer 2: ```suggestion
Compose YAML into a single C file
+++++++++++++++++++++++++++++++++
```

> Now all necessary information is saved in ``mpm88.yml``. However, the ``kernel_source``'s are separated one-by-one.

Reviewer: ```suggestion
Now, all necessary information is saved in ``mpm88.yml``, in the form of multiple separate records.
```

> Now all necessary information is saved in ``mpm88.yml``. However, the ``kernel_source``s are separated one-by-one.
> You may want to **compose** the separate kernels into **one single file** in order to distribute it easier.

Reviewer 2: ```suggestion
Now all necessary information is saved in ``mpm88.yml``. However, the ``kernel_source`` s are separated one-by-one.
You may want to **compose** the separate kernels into **one single file** in order to distribute it more easily.
```

> You may want to **compose** the separate kernels into **one single file** in order to distribute it more easily.

Reviewer: ```suggestion
You may want to **compose** the separate kernels into **one single file** for more portability.
```

> We provide a useful tool to do this, type these commands to your console:

Reviewer 2: ```suggestion
We provide a useful tool to do this, type these commands in your console:
```

> We provide a useful tool to do this, type these commands in your console:

Reviewer: ```suggestion
We provide a useful CLI tool to do this:
```

.. code-block:: bash

   python3 -m taichi cc_compose mpm88.yml mpm88.c mpm88.h

This composes all the kernels and runtimes in ``mpm88.yml`` into a single C
source file ``mpm88.c``:

.. code-block:: c

    ...

    Ti_i8 Ti_gtmp[1048576];
    union Ti_BitCast Ti_args[8];
    Ti_i32 Ti_earg[8 * 8];

    struct Ti_Context Ti_ctx = {  // statically-allocated context for convenience!
      &Ti_root, Ti_gtmp, Ti_args, Ti_earg,
    };

    void Tk_init_c6_0(struct Ti_Context *ti_ctx) {
      for (Ti_i32 tmp0 = 0; tmp0 < 8192; tmp0 += 1) {
        Ti_i32 tmp1 = tmp0;
        Ti_f32 tmp2 = Ti_rand_f32();
        Ti_f32 tmp3 = Ti_rand_f32();
        Ti_f32 tmp4 = 0.4;
        Ti_f32 tmp5 = tmp2 * tmp4;

        ...

> And a C header file ``mpm88.h`` for declarations of data structures, functions
> (kernels) for this file.

Reviewer: ```suggestion
... and a C header file ``mpm88.h`` for declarations of data structures, functions
(Taichi kernels) for this file.
```

.. note::

   The generated C source is promised to be C99 compatible.

>    It should be functional when being compiled as C++ as well.

Reviewer 2: ```suggestion
   It should also be functional when being compiled as C++.
```

>    The generated C source is promised to be C99 compatible.
> 
>    It should also be functional when being compiled as C++.

Reviewer: ```suggestion
   The generated C source is promised to be C99 compatible.

   It should also be functional when compiled using a C++ compiler.
```

Calling the exported kernels
----------------------------

> Then, link this file (``mpm88.c``) together with your C/C++ project.
> Include the header (``mpm88.h``) to where kernels are to be called.

Reviewer: ```suggestion
Then, link the C file (``mpm88.c``) against your C/C++ project.
Include the header file (``mpm88.h``) when Taichi kernels are called.
```

> To call the kernel ``init_c6_0``, for example:

Reviewer: ```suggestion
For example, calling kernel ``init_c6_0`` can be implemented as follows:
```

.. code-block:: cpp

    #include "mpm88.h"

    int main(void) {
        ...
        Tk_init_c6_0(&Ti_ctx);
        ...
    }

> Or, if you need multiple Taichi context within one program:

Reviewer 2: ```suggestion
Alternatively, if you need multiple Taichi contexts within one program:
```

      // allocate buffers on your own:
      per_renderer_taichi_context.root = malloc(...);
      ...
      Tk_init_c6_0(&per_renderer_taichi_context);
    }

Specifying scalar arguments
+++++++++++++++++++++++++++

To specify scalar arguments for kernels:

.. code-block:: cpp

    Ti_ctx.args[0].val_f64 = 3.14;  // first argument, float64
    Ti_ctx.args[1].val_i32 = 233;  // second argument, int32
    Tk_my_kernel_c8_0(&Ti_ctx);
    double ret = Ti_ctx.args[0].val_f64;  // return value, float64

    printf("my_kernel(3.14, 233) = %lf\n", ret);

Passing external arrays
+++++++++++++++++++++++

To pass external arrays as arguments for kernels:

.. code-block:: cpp

    float img[640 * 480 * 3];

    Ti_ctx.args[0].ptr_f32 = img;  // first argument, float32 pointer to array

    // specify the shape of that array:
    Ti_ctx.earg[0 * 8 + 0] = 640;  // img.shape[0]
    Ti_ctx.earg[0 * 8 + 1] = 480;  // img.shape[1]
    Ti_ctx.earg[0 * 8 + 2] = 3;    // img.shape[2]
    Tk_matrix_to_ext_arr_c12_0(&Ti_ctx);

    // note that the array used in Taichi is row-major:
    printf("img[3, 2, 1] = %f\n", img[(3 * 480 + 2) * 3 + 1]);

Taichi.js (WIP)
---------------

> Once you have C source generated, you can compile the C source into Javascript

Reviewer: ```suggestion
Once you have C source file generated, you can compile them into Javascript
```

or WASM via Emscripten.

We provide `Taichi.js <https://github.com/taichi-dev/taichi.js>`_ as an
infrastructure for wrapping Taichi kernels for Javascript.
See `its README.md <https://github.com/taichi-dev/taichi.js/blob/master/README.md>`_ for the complete workflow.

> Check `this page <https://taichi-dev.github.io/taichi.js>`_ for online demo.

Reviewer: ```suggestion
Check out `this page <https://taichi-dev.github.io/taichi.js>`_ for online demos.
```

Calling Taichi kernels from Julia (WIP)
---------------------------------------

Once you have C source generated, you can then compile the C source into a
shared object. Then it can be called from other langurages that provides a C
interface, including but not limited to Julia, Matlab, Mathematica, Java, etc.

TODO: WIP.

> 
> The export workflow for Metal shaders (WIP)
> -------------------------------------------
> 
> Actually we also support exporting / dumping Metal shaders for the Apple Metal
> backend via the shared API as C backend does. Documentation WIP.

Reviewer: ```suggestion
```

Sorry, but we cannot mention this yet due to a nondisclosure agreement.

Export Taichi kernels to C source
=================================

The C backend of Taichi allows you to **export Taichi kernels to C source**.

The exported Taichi program which is simply consist of C99 compatible sources
does not necessary to be launched from Python. Instead, you may use them in

> your C/C++ project. Or even calling them in Javascript via Empscripten.

Reviewer 3: ```suggestion
The exported Taichi program consists purely of C99 compatible sources and does not require Python. Instead, you can use the exported sources in a C/C++ project, or even by calling them in Javascript via Empscripten.
 ```

> The exported Taichi program which is simply consist of C99 compatible sources
> does not necessary to be launched from Python. Instead, you may use them in
> your C/C++ project. Or even calling them in Javascript via Empscripten.

Author: ```suggestion
The exported Taichi program consists purely of C99 compatible sources and does not require Python. Instead, you can use the exported sources in a C/C++ project, or even by calling them in Javascript via Empscripten.
```

Each C function corresponds to a Taichi kernel.
For example, ``Tk_init_c6_0()`` may corresponds to ``init()`` in ``mpm88.py``.

Only the source is needed when copying, all the required Taichi runtimes are
included in that single C source file.

> This also allows commercial people to distribute their Taichi program in

Reviewer 3: ```suggestion
For example, this also allows commercial users to distribute their Taichi program in
```

binary format by linking this file with their project.

.. warning::

>     Currently, this feature is only officially supported on the C backend which is only released on **Linux** platform at this moment.

Reviewer 3: ```suggestion
    Currently, this feature is only officially supported on the C backend which is only released on **Linux** at this moment.
```

Export Taichi kernels to C source
=================================

The C backend of Taichi allows you to **export Taichi kernels to C source**.

> The exported Taichi program, which are simply C99 compatible source,

Reviewer 4: ```suggestion
The exported Taichi program which is simply consist of C99 compatible sources
```

does not necessary to be launched from Python. Instead, you may use them in
your C/C++ project. Or even calling them in Javascript via Empscripten.

Each C function corresponds to a Taichi kernel.
For example, ``Tk_init_c6_0()`` may corresponds to ``init()`` in ``mpm88.py``.

Only the source is need when copying, all the required Taichi runtimes are

> included in that single C source file.

Reviewer 4: ```
Only the source is needed when copying, all the required Taichi runtimes are
included in that single C source file.
```
Btw, sorry for my bad understanding, but I feel a little hard to catch the point :(
What is `copying` here? Do we need to clarify it slightly?

This also allows commercial people to distribute their Taichi program in
binary format by linking this file with their project.

> .. note::

Reviewer 4: ```suggestion
.. warning::
```
Will this be more obvious?

>     Currently this feature is only officially supported on the C backend.
>     Also note that the C backend is only released on **Linux** platform for now.

Reviewer 4: ```suggestion
    Currently, this feature is only officially supported on the C backend which is only released on **Linux** platform at this moment.
```

> The workflow of export

Reviewer 4: ```suggestion
The workflow of exporting
```

----------------------

Use ``ti.core.start_recording`` in the Taichi program you want to export.

Suppose you want to export the `examples/mpm88.py <https://github.com/taichi-dev/taichi/blob/master/examples/mpm88.py>`_, here are the workflow:

Export YAML
+++++++++++

> First, modify the ``mpm88.py`` a little bit:

Reviewer 4: ```suggestion
First, modifying the ``mpm88.py`` as shown below:
```

.. code-block:: python

    import taichi as ti

    ti.core.start_recording('mpm88.yml')
    ti.init(arch=ti.cc)

    ... # your program

And run it. Close the GUI window once particles are shown up correctly.

This will save all the kernels in ``mpm88.py`` to ``mpm88.yml``:

.. code-block:: yaml

   - action: "compile_kernel"
      kernel_name: "init_c6_0"
      kernel_source: "void Tk_init_c6_0(struct Ti_Context *ti_ctx) {\n  for (Ti_i32 tmp0 = 0; tmp0 < 8192...\n"
    - action: "launch_kernel"
      kernel_name: "init_c6_0"
    ...

.. note::

    Equivalently, you may also specify these two arguments from environment

>     variables if you are on Unix-alike system:

Reviewer 4: ```suggestion
    variables on Unix-like system:
```
Unix-like is a better terminology, as (!wiki)[https://en.wikipedia.org/wiki/Unix-like] showed IMO

    .. code-block:: bash

        TI_ARCH=cc TI_ACTION_RECORD=mpm88.yml python mpm88.py

> Compose YAML into single C file

Reviewer 4: ```suggestion
Compose YAML into a single C file
```

+++++++++++++++++++++++++++++++

> Now all necessary information are saved in ``mpm88.yml``.
> However the ``kernel_source``'s are separated one-by-one.
> So you may want to **compose** the separate kernels into **one single file**
> to make distribution easier.

Reviewer 4: ```suggestion
Now all necessary information is saved in ``mpm88.yml``. However, the ``kernel_source``s are separated one-by-one.
You may want to **compose** the separate kernels into **one single file** in order to distribute it easier.
```

> We provide a useful tool to compose them, type these commands to your console:

Reviewer 4: ```suggestion
We provide a useful tool to do this, type these commands to your console:
```
Because `compose` has been used above
